// @storybook/vue3/dist/_node-chunks/parsers-YR2XJJNW.js
import CJS_COMPAT_NODE_URL_xt2tpo40odg from 'node:url';
import CJS_COMPAT_NODE_PATH_xt2tpo40odg from 'node:path';
import CJS_COMPAT_NODE_MODULE_xt2tpo40odg from "node:module";

var __filename = CJS_COMPAT_NODE_URL_xt2tpo40odg.fileURLToPath(import.meta.url);
var __dirname = CJS_COMPAT_NODE_PATH_xt2tpo40odg.dirname(__filename);
var require = CJS_COMPAT_NODE_MODULE_xt2tpo40odg.createRequire(import.meta.url);

// ------------------------------------------------------------
// end of CJS compatibility banner, injected by Storybook's esbuild configuration
// ------------------------------------------------------------

// src/parsers/vueImportParser.ts
import {
  ChangeDetectionFailureError
} from "storybook/internal/core-server";
var compilerSfcPromise;
async function getCompilerSfc() {
  return compilerSfcPromise || (compilerSfcPromise = import("vue/compiler-sfc").catch((error) => {
    throw compilerSfcPromise = void 0, new ChangeDetectionFailureError(
      `Failed to load 'vue/compiler-sfc'. Original error: ${error instanceof Error ? error.message : String(error)}`,
      error instanceof Error ? { cause: error } : void 0
    );
  })), compilerSfcPromise;
}
function virtualExtensionForLang(lang) {
  switch (lang) {
    case "ts":
      return "ts";
    case "tsx":
      return "tsx";
    case "jsx":
      return "jsx";
    default:
      return "js";
  }
}
var vueImportParser = {
  extensions: [".vue"],
  async parse({ filePath, source }, ctx) {
    let sfc = await getCompilerSfc(), parsed;
    try {
      parsed = sfc.parse(source, { filename: filePath });
    } catch (error) {
      throw new ChangeDetectionFailureError(
        `vue/compiler-sfc failed to parse ${filePath}: ${error instanceof Error ? error.message : String(error)}`,
        error instanceof Error ? { cause: error } : void 0
      );
    }
    if (parsed.errors.length > 0) {
      let first = parsed.errors[0], message = first instanceof Error ? first.message : String(first);
      throw new ChangeDetectionFailureError(
        `vue/compiler-sfc reported errors for ${filePath}: ${message}`
      );
    }
    let { descriptor } = parsed, blocks = [];
    if (descriptor.script && descriptor.script.content.length > 0 && blocks.push({ content: descriptor.script.content, lang: descriptor.script.lang }), descriptor.scriptSetup && descriptor.scriptSetup.content.length > 0 && blocks.push({
      content: descriptor.scriptSetup.content,
      lang: descriptor.scriptSetup.lang
    }), blocks.length === 0)
      return [];
    let edges = [], seen = /* @__PURE__ */ new Set();
    for (let block of blocks) {
      let ext = virtualExtensionForLang(block.lang), virtualFilePath = `${filePath}.script.${ext}`, scriptEdges = await ctx.parseScriptWithOxc(block.content, virtualFilePath);
      for (let edge of scriptEdges) {
        let key = `${edge.kind}:${edge.specifier}`;
        seen.has(key) || (seen.add(key), edges.push(edge));
      }
    }
    return edges;
  }
};
export {
  vueImportParser
};
